File: leetcode/binary-tree-maximum-path-sum/binary-tree-maximum-path-sum.cpp
/**************************************
* @author dujiong
* @date 2017.1.5
* @version V0.1
**************************************/

/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    int maxPathSum(TreeNode *root) {
        int ans = INT_MIN;
        dfs(root, ans);
        return ans;
    }
private:
    int dfs(TreeNode* root, int& ans){
        if(root==NULL) return 0;
        int l = std::max(0, dfs(root->left, ans));
        int r = std::max(0, dfs(root->right, ans));
        ans = std::max(ans, l+r+root->val);
        return std::max(l, r)+root->val;
    }
};

File: leetcode-golang/Binary Tree Maximum Path Sum/Binary Tree Maximum Path Sum.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
 func maxPathSum(root *TreeNode) int {
    ans := math.MinInt32
    dfs(root, &ans)
    return ans
}

func dfs(root *TreeNode, ans *int) int {
    if root == nil {
        return 0
    }
    left := max(0, dfs(root.Left, ans))
    right := max(0, dfs(root.Right, ans))
    *ans = max(*ans, left + right + root.Val)
    return max(left, right) + root.Val
}

func max(i, j int) int {
    if i > j {
        return i
    }
    return j
}